x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -657,6 +660,86 @@ public void getUserPrivileges(AuthorizationInfo authorizationInfo, ActionListene
         }
     }
 
+    @Override
+    public void getRemoteAccessRoleDescriptorsIntersection(
+        final String remoteClusterAlias,
+        final AuthorizationInfo authorizationInfo,
+        final ActionListener<RoleDescriptorsIntersection> listener
+    ) {
+        if (authorizationInfo instanceof RBACAuthorizationInfo == false) {
+            listener.onFailure(
+                new IllegalArgumentException("unsupported authorization info:" + authorizationInfo.getClass().getSimpleName())
+            );
+            return;
+        }
+
+        final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+        final RemoteIndicesPermission remoteIndicesPermission;
+        try {
+            remoteIndicesPermission = role.remoteIndices().forCluster(remoteClusterAlias);
+        } catch (UnsupportedOperationException e) {
+            // TODO we will need to implement this to support API keys with assigned role descriptors
+            final String message = "Cannot retrieve remote access role descriptors for API keys with assigned role descriptors.";
+            assert false : message;
+            listener.onFailure(new IllegalArgumentException(message, e));
+            return;
+        }
+
+        if (remoteIndicesPermission.remoteIndicesGroups().isEmpty()) {
+            listener.onResponse(RoleDescriptorsIntersection.EMPTY);
+            return;
+        }
+
+        final List<RoleDescriptor.IndicesPrivileges> indicesPrivileges = new ArrayList<>();
+        for (RemoteIndicesPermission.RemoteIndicesGroup remoteIndicesGroup : remoteIndicesPermission.remoteIndicesGroups()) {
+            for (IndicesPermission.Group indicesGroup : remoteIndicesGroup.indicesPermissionGroups()) {
+                indicesPrivileges.add(toIndicesPrivileges(indicesGroup));
+            }
+        }
+
+        listener.onResponse(
+            new RoleDescriptorsIntersection(
+                List.of(
+                    Set.of(
+                        new RoleDescriptor(
+                            UUIDs.randomBase64UUID(),
```

Comment:
I think we should not use random ID here since it does not work with caching (assuming cache key depends on the raw bytes of RoleDescriptor). I don't mind to use a constant here since it is really just a placeholder. As long as the bytes of the actual content are the same, two role descriptors should be able to share the cache entry. This is makes it possible to share the same role for requests from different users and they can even be from different clusters (as long as the role descriptor content is identical).

Reply:
Yes ofc, this shouldn't be random. 

I think we might want to be careful though in sharing the cache across clusters, as it provides a convenient timing attack vector for one querying cluster to learn about the privileges of another cluster's users. We can easily prevent this though by prefixing a unique per-querying-cluster entry to the cache key on FC; let's revisit this discussion when we get to the FC implementation -- there's security/performance trade-offs here. 

Regardless of whether we share the cache across QCs, the RDs sent here should be deterministic and having a random identifier makes no sense.